### PR TITLE
Blocks and snapshot compression

### PR DESCRIPTION
**Pull request by keorn**
- [x] reduce `Compressible` interface
- [x] check migrations
- [x] benchmark effect of `blocks` DB compression on syncing

### CODE REVIEW

File: util/src/rlp/rlpcompression.rs [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+	let to_swap: &[&[u8]] = &[&[0x83, b'c', b'a', b't'], &[0x83, b'd', b'o', b'g']];
+	let invalid_rlp: &[&[u8]] = &[&[0x81, 0x00], &[0x81, 0x01]];
+	let swapper = InvalidRlpSwapper::new(to_swap, invalid_rlp);
+	assert_eq!(Some(invalid_rlp[0]), swapper.get_invalid(&[0x83, b'c', b'a', b't']));
+	assert_eq!(None, swapper.get_invalid(&[0x83, b'b', b'a', b't']));
+	assert_eq!(Some(to_swap[1]), swapper.get_valid(invalid_rlp[1]));
+}
+
+/// Type of RLP indicating its origin database.
+pub enum RlpType {
```
Comment (keorn) [inline util/src/rlp/rlpcompression.rs]: I don't like this, should it be a field of `UntrustedRlp` which indicates origin.

File: util/src/rlp/rlpcompression.rs [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+		self.invalid_to_valid.get(invalid_rlp).map(|r| r.clone())
+	}
+	/// Get an invalid RLP corresponding to a valid one
+	fn get_invalid(&self, valid_rlp: &[u8]) -> Option<&[u8]> {
+		self.valid_to_invalid.get(valid_rlp).map(|r| r.clone())
+	}
+}
+
+#[test]
+fn invalid_rlp_swapper() {
```
Comment (rphmeier) [inline util/src/rlp/rlpcompression.rs]: tests should be in their own `#[cfg(test)]` submodule so we don't compile test functions into a non-test build

**Comment by keorn** [comment]
Around 4% `import` slowdown on first 200k blocks.

**Comment by keorn** [comment]
Blocks DB goes from 227MiB to 110MiB looking at raw bytes at block 324k.

File: ethcore/src/account.rs [hunk 3]
```diff
@@ -289,6 +289,16 @@ mod tests {
...
 	#[test]
+	fn account_compress() {
+    let raw = Account::new_basic(2.into(), 4.into()).rlp();
+    let rlp = UntrustedRlp::new(&raw);
+    let compact_vec = rlp.compress(RlpType::Snapshot).to_vec();
+    assert!(raw.len() > compact_vec.len());
+    let again_raw = UntrustedRlp::new(&compact_vec).decompress(RlpType::Snapshot);
+    assert_eq!(raw, again_raw.to_vec());
+  }
+
```
Comment (debris) [inline ethcore/src/account.rs]: replace spaces with tabs

File: util/src/rlp/commonrlps.rs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Contains common RLPs used for compression found using the following test.
+
+use rlp::rlpcompression::InvalidRlpSwapper;
+
+#[test]
+#[ignore]
+fn analyze_db() {
```
Comment (debris) [inline util/src/rlp/commonrlps.rs]: to be consistent with other source files, it's be better to put `#[test]` function at the end of the file

Comment (rphmeier) [inline util/src/rlp/commonrlps.rs]: they should also be marked `#[cfg(test)]` explicitly or within a module marked as such

Comment (keorn) [inline util/src/rlp/commonrlps.rs]: yep, will need to move other ones throughout the codebase too.

File: util/src/rlp/rlpcompression.rs [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+fn deep_compress(rlp: &UntrustedRlp, swapper: &InvalidRlpSwapper) -> Option<ElasticArray1024<u8>> {
+	let simple_swap = ||
+		swapper.get_invalid(rlp.as_raw()).map(|b| to_elastic(&b));
+	if rlp.is_data() {
+		// Try to treat the inside as RLP.
+		return match rlp.payload_info() {
+			// Shortest decompressed account is 70, so simply try to swap the value.
+			Ok(ref p) if p.value_len < 70 => simple_swap(),
+			_ => {
+				if let Ok(d) = rlp.data() {
```
Comment (debris) [inline util/src/rlp/rlpcompression.rs]: line `109` already proves that rlp is data. You can use `expect(...)` here or replace line 109 with:

`if let Ok(ref d)  = rlp.data() {`

File: util/src/rlp/rlpcompression.rs [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+					let internal_rlp = UntrustedRlp::new(&d);
```
Comment (debris) [inline util/src/rlp/rlpcompression.rs]: redundant `&d` here, `rlp.data()` returns a reference

File: util/src/rlp/rlpcompression.rs [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+	// Simply decompress data.
+	if rlp.is_data() { return simple_swap(); }
+	match rlp.item_count() {
+		// Look for special compressed list, which contains nested data.
+		2 if rlp.at(0).map(|r| r.as_raw() == &[0x81, 0x7f]).unwrap_or(false) =>
+			rlp.at(1).ok().map_or(simple_swap(),
+			|r| deep_decompress(&r, swapper).map(|d| { let v = d.to_vec(); encode(&v) })),
+		// Iterate through RLP while checking if it has been compressed.
+		_ => map_rlp(rlp, |r| deep_decompress(r, swapper)),
+  	}
```
Comment (debris) [inline util/src/rlp/rlpcompression.rs]: you can avoid quick return in line `139` by using `prototype()` method.

```rust
match rlp.prototype() {
	Prototype::Data(_) => simple_swap(),
	Prototype::List(item_count) if /* guard */ => /* your logic */,
	_ => /* else logic */
}
```

File: util/src/rlp/rlpcompression.rs (hunk 5 above)
Comment (debris) [inline util/src/rlp/rlpcompression.rs]: *edit, my previous comment may not be valid. `is_data` does not check validity of data, whereas `rlp.data()` does. Is `rlp` expected to be invalid here?

File: util/src/rlp/rlpcompression.rs (hunk 7 above)
Comment (debris) [inline util/src/rlp/rlpcompression.rs]: *edit, same as before. My comment is probably not valid, cause you are working on malformed data.

**Comment by debris** [comment]
overall LGTM

File: util/src/rlp/rlpcompression.rs (hunk 5 above)
Comment (keorn) [inline util/src/rlp/rlpcompression.rs]: It can be invalid, since data field is not always an RLP.

File: util/src/rlp/rlpcompression.rs (hunk 7 above)
Comment (keorn) [inline util/src/rlp/rlpcompression.rs]: Yes

**Comment by gavofyork** [comment]
i guess it's not yet optional?

**Comment by keorn** [comment]
Ah ok, will make it optional for `blocks`.

**Comment by keorn** [comment]
Waiting for #1600.

**Comment by gavofyork** [comment]
ok looks good. actually results in a ~ 8% speedup for the first 200k blocks.

**Comment by keorn** [comment]
I got some speedup on full sync too.
Master:
```
real    405m49.204s
user    587m33.366s
sys     27m45.241s
```
With blocks compression:
```
real    395m6.620s
user    557m51.039s
sys     25m58.350s
```

**Pull request merged**